File: Framework/Php/framework/framework/template/smarttemplate.php
<?php

/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */

namespace Smartkernel\Framework\Template {

    use Smartkernel\Framework\SmartString;
    use Smartkernel\Framework\Text\SmartEncoding;
    use Smartkernel\Framework\IO\SmartFile;

/*
     * 模板生成器
     */

    class SmartTemplate {
        /*
         * 生成
         * 
         * @param string $templet，模板
         * @param array $data，参数
         * @return string，结果
         */

        public static function generate($templet, array $data = null) {
            if ($data != null) {
                foreach ($data as $key => $value) {
                    $templet = SmartString::repeatReplace($templet, SmartString::format("@({0})", $key), $value);
                }
            }
            return $templet;
        }

        /*
         * 生成
         * 
         * @param string $templet，模板
         * @param string $filePath，保存路径
         * @param array $data，参数
         * @param bool $isOverWrite，是否覆盖
         * @param string $encoding，文件编码
         * @param string $filePathEncoding，文件路径编码
         * @return string，结果
         */

        public static function generateFile($templet, $filePath, array $data = null, $isOverWrite = false, $encoding = null, $filePathEncoding = SmartEncoding::DefaultSafePathValue) {
            $encoding = $encoding == null ? SmartEncoding::DefaultValue : $encoding;

            $code = self::generate($templet, $data);
            if ($isOverWrite) {
                //覆盖
                SmartFile::delete($filePath, $filePathEncoding);
                SmartFile::write($filePath, $code, true, $encoding, $filePathEncoding);
            } else {
                //不覆盖
                if (!SmartFile::exists($filePath, $filePathEncoding)) {
                    SmartFile::write($filePath, $code, true, $encoding, $filePathEncoding);
                }
            }
        }

    }

}

File: Framework/CSharp/Framework/Framework/Template/SmartTemplate.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using Smartkernel.Framework.IO;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Smartkernel.Framework.Template
{
    /// <summary>
    /// 模板生成器
    /// </summary>
    public class SmartTemplate
    {
        /// <summary>
        /// 生成
        /// </summary>
        /// <param name="templet">模板</param>
        /// <param name="data">参数</param>
        /// <param name="imports">命名空间</param>
        /// <param name="references">程序集引用</param>
        /// <returns>结果</returns>
        public static string Generate(string templet, Dictionary<string, object> data = null, List<string> imports = null, List<string> references = null)
        {
            SmartEngine engine = new SmartEngine();
            SmartEngineHost engineHost = new SmartEngineHost(templet);

            if (imports != null && imports.Count > 0)
            {
                imports.ForEach(item =>
                {
                    engineHost.StandardImports.Add(item);
                });
            }
            if (references != null && references.Count > 0)
            {
                references.ForEach(item =>
                {
                    engineHost.StandardAssemblyReferences.Add(item);
                });
            }

            if (data != null)
            {
                foreach (var kv in data)
                {
                    var key = kv.Key;
                    var value = kv.Value;
                    engineHost.Data.Add(key, value);
                }
            }
            string output = engine.Process(engineHost);
            string errors = string.Empty;
            for (int j = 0; j < engineHost.Errors.Count; j++)
            {
                errors = errors + @"\r" + engineHost.Errors[j];
            }
            if (!string.IsNullOrWhiteSpace(errors))
            {
                throw new Exception(errors);
            }
            else
            {
                return output;
            }
        }

        /// <summary>
        /// 生成
        /// </summary>
        /// <param name="templet">模板</param>
        /// <param name="data">参数</param>
        /// <param name="filePath">保存路径</param>
        /// <param name="isOverWrite">是否覆盖</param>
        /// <param name="encoding">编码</param>
        /// <param name="imports">命名空间</param>
        /// <param name="references">程序集引用</param>
        public static void GenerateFile(string templet, string filePath, Dictionary<string, object> data = null, bool isOverWrite = false, Encoding encoding = null, List<string> imports = null, List<string> references = null)
        {
            encoding = encoding ?? Encoding.Unicode;//默认的编码格式会在一行上，只有Unicode才是多行，这样避免代码行数统计错误

            var code = Generate(templet, data);
            if (isOverWrite)
            {
                //覆盖
                File.Delete(filePath);
                SmartFile.Write(filePath, code, true, encoding);
            }
            else
            {
                //不覆盖
                if (!File.Exists(filePath))
                {
                    SmartFile.Write(filePath, code, true, encoding);
                }
            }
        }
    }
}
